clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -296,31 +254,78 @@ public void send(ClientRequest request, long now) {
-        ClientRequest clientRequest = new ClientRequest(nodeConnectionId, now, true,
-                nextRequestHeader(ApiKeys.METADATA), metadataRequest, null);
+    private void sendInternalMetadataRequest(MetadataRequest.Builder builder,
+                                             String nodeConnectionId, long now) {
+        ClientRequest clientRequest = newClientRequest(
+                nodeConnectionId, builder, now, true, null);
         doSend(clientRequest, true, now);
     }
 
-    private void doSend(ClientRequest request, boolean isInternalRequest, long now) {
-        String nodeId = request.destination();
-        if (request.header().apiKey() == ApiKeys.API_VERSIONS.id) {
-            if (!canSendApiVersionsRequest(nodeId))
-                throw new IllegalStateException("Attempt to send API Versions request to node " + nodeId + " which is not ready.");
-        } else if (!canSendRequest(nodeId))
-            throw new IllegalStateException("Attempt to send a request to node " + nodeId + " which is not ready.");
-
-        Send send = request.body().toSend(nodeId, request.header());
+    private void doSend(ClientRequest clientRequest, boolean isInternalRequest, long now) {
+        String nodeId = clientRequest.destination();
+        if (!isInternalRequest) {
+            // If this request came from outside the NetworkClient, validate
+            // that we can send data.  If the request is internal, we trust
+            // that that internal code has done this validation.  Validation
+            // will be slightly different for some internal requests (for
+            // example, ApiVersionsRequests can be sent prior to being in
+            // READY state.)
+            if (!canSendRequest(nodeId))
+                throw new IllegalStateException("Attempt to send a request to node " + nodeId + " which is not ready.");
+        }
+        AbstractRequest request = null;
+        AbstractRequest.Builder builder = clientRequest.requestBuilder();
+        try {
+            NodeApiVersions versionInfo = nodeApiVersions.get(Integer.parseInt(nodeId));
+            // Note: if versionInfo is null, we have no server version information.
+            // This would be the case when sending the initial ApiVersionRequest which
+            // fetches the version information itself, for example.
+            if (versionInfo == null) {
+                if (log.isTraceEnabled()) {
+                    log.trace("No version information found when sending message of type {} to node {}",
+                            clientRequest.apiKey(), nodeId);
+                }
+            } else {
+                short version = versionInfo.getUsableVersion(clientRequest.apiKey().id);
+                if (log.isTraceEnabled()) {
+                    log.trace("When sending message of type {} to node {}, the best usable " +
+                            "version is {}", clientRequest.apiKey(), nodeId, version);
```

Comment:
It would be good to include the most current version in this message as well. Also, as Jun said it may be a good to log at a higher level if we are downgrading a request. Not sure what Jun had in mind, but maybe `debug`?